### PR TITLE
add boltons

### PR DESCRIPTION
**Pull request by AABur**
add stubs for [boltons](https://github.com/mahmoud/boltons) library

### CODE REVIEW

**Comment by AABur** [comment]
@sobolevn
Please review the bolton/dictutils stubs
This is my first PR in typeshed and there is for sure room for improvement.

File: stubs/boltons/boltons/cacheutils.pyi [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    bound_to: Incomplete
+    def __init__(self, func, cache, scoped: bool = ..., typed: bool = ..., key: Incomplete | None = ...): ...
+    def __get__(self, obj, objtype: Incomplete | None = ...): ...
+    def __call__(self, *args, **kwargs): ...
+
+def cached(cache, scoped: bool = ..., typed: bool = ..., key: Incomplete | None = ...): ...
+def cachedmethod(cache, scoped: bool = ..., typed: bool = ..., key: Incomplete | None = ...): ...
+
+class cachedproperty:
+    __doc__: Incomplete
```
Comment (sobolevn) [inline stubs/boltons/boltons/cacheutils.pyi:72]: Can be removed.

File: stubs/boltons/boltons/iterutils.pyi [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+class GUIDerator:
+    size: Incomplete
+    count: Incomplete
+    def __init__(self, size: int = ...) -> None: ...
+    pid: Incomplete
+    salt: Incomplete
+    def reseed(self) -> None: ...
+    def __iter__(self): ...
+    def __next__(self): ...
+    def __next__(self): ...
```
Comment (sobolevn) [inline stubs/boltons/boltons/iterutils.pyi:66]:
```suggestion
    def __next__(self): ...
```

Defined twice.

File: stubs/boltons/boltons/iterutils.pyi [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    next: Incomplete
+
+class SequentialGUIDerator(GUIDerator):
+    start: Incomplete
+    def reseed(self) -> None: ...
+    def reseed(self) -> None: ...
```
Comment (sobolevn) [inline stubs/boltons/boltons/iterutils.pyi:72]:
```suggestion
    def reseed(self) -> None: ...
```

Defined twice.

File: stubs/boltons/boltons/tbutils.pyi [hunk 4]
```diff
@@ -0,0 +1,91 @@
+from _typeshed import Incomplete
+
+text = unicode
```
Comment (sobolevn) [inline stubs/boltons/boltons/tbutils.pyi:3]: `unicode` is gone.
```suggestion
```

**Review by sobolevn — DISMISSED** [review]
Thanks, this is good start.

I think we would need to:
1. Fix flake8. Right now it says:

```
 ./stubs/boltons/boltons/timeutils.pyi:46:20: F821 undefined name 'DSTEND_1987_2006'
./stubs/boltons/boltons/funcutils.pyi:45:5: F811 redefinition of unused 'get_sig_str' from line 43
./stubs/boltons/boltons/funcutils.pyi:46:5: F811 redefinition of unused 'get_invocation_str' from line 44
./stubs/boltons/boltons/funcutils.pyi:54:5: F811 redefinition of unused 'add_arg' from line 53
./stubs/boltons/boltons/urlutils.pyi:92:1: F811 redefinition of unused 'OrderedMultiDict' from line 5
./stubs/boltons/boltons/deprutils.pyi:5:26: F821 undefined name 'ModuleType'
./stubs/boltons/boltons/ioutils.pyi:5:13: F821 undefined name 'unicode'
./stubs/boltons/boltons/tbutils.pyi:3:8: F821 undefined name 'unicode'
./stubs/boltons/boltons/fileutils.pyi:50:12: F821 undefined name 'copy_tree'
./stubs/boltons/boltons/iterutils.pyi:66:5: F811 redefinition of unused '__next__' from line 65
./stubs/boltons/boltons/iterutils.pyi:72:5: F811 redefinition of unused 'reseed' from line 71
```

You need to define / import all missing names.

2. Mypy and other tools must pass, there are multiple issues there at the moment. For example,

```
stubs/boltons/boltons/listutils.pyi:26: note:      Superclass:
stubs/boltons/boltons/listutils.pyi:26: note:          @overload
stubs/boltons/boltons/listutils.pyi:26: note:          def sort(self, *, key: None = ..., reverse: bool = ...) -> None
stubs/boltons/boltons/listutils.pyi:26: note:          @overload
stubs/boltons/boltons/listutils.pyi:26: note:          def sort(self, *, key: Callable[[Any], Union[SupportsDunderLT[Any], SupportsDunderGT[Any]]], reverse: bool = ...) -> None
stubs/boltons/boltons/listutils.pyi:26: note:      Subclass:
stubs/boltons/boltons/listutils.pyi:26: note:          def sort(self) -> None
```

The only way to fix it is to add `# type: ignore` to the method definition line: `def sort(self) -> None:  # type: ignore`

**Comment by AABur** [comment]
flake8 & mypy FIXED

```
--- TEST SUMMARY: One or more tests failed. See above for details. ---

Flake8: Success
Check consistent: Success
Check new syntax: Success
Pyright: Success
mypy: Success
stubtest: Failed
pytype: Success
Regression test: Success
```

Time to dive into designing the correct stubs

**Comment by AlexWaygood** [comment]
Thanks @AABur. Are you able to take a look at the stubtest errors in CI? Please say if you need guidance on any of them :)

File: stubs/boltons/@tests/stubtest_allowlist.txt [hunk 5]
```diff
@@ -0,0 +1,3 @@
+boltons.ecoutils
```
Comment (AlexWaygood) [inline stubs/boltons/@tests/stubtest_allowlist.txt:1]:
```suggestion
```

**Comment by AABur** [comment]
stubtest don't working in my local environment

```sh
❯ py tests/stubtest_third_party.py boltons
boltons... skipping, unsupported platform: darwin, supported: ['linux']
```

Help me to fix it!

**Comment by AlexWaygood** [comment]
> stubtest don't working in my local environment
> 
> ```shell
> ❯ py tests/stubtest_third_party.py boltons
> boltons... skipping, unsupported platform: darwin, supported: ['linux']
> ```
> 
> Help me to fix it!

It'll be fixed by #9173. Sorry for the inconvenience!

**Comment by sobolevn** [comment]
Stubtest only has issues with `boltons.ecoutils.*` constants 🎉

**Comment by AABur** [comment]
`pyright` and `stubtest` errors are not detected when tested locally on a Mac

File: stubs/boltons/boltons/ecoutils.pyi [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+RLIMIT_FDS_SOFT: int
+RLIMIT_FDS_HARD: int
+START_TIME_INFO: dict[str, str | float]
+
+def get_python_info() -> dict[str, Any]: ...
+def get_profile(**kwargs) -> dict[str, Any]: ...
+def get_profile_json(indent: bool = ...) -> str: ...
+def main() -> NoReturn: ...
+
+# def dumps(val: Any, indent: int) -> str: ...
```
Comment (sobolevn) [inline stubs/boltons/boltons/ecoutils.pyi:31]: Looks like this does exist: https://github.com/mahmoud/boltons/blob/6c7fa3e05659fac6515f661153c89d4ad02df168/boltons/ecoutils.py#L360-L363

Comment (AABur) [reply stubs/boltons/boltons/ecoutils.pyi:31]: Yes, there exists.
But `stubtest` complains about it in runtime https://github.com/python/typeshed/actions/runs/3523498827/jobs/5907729287#step:5:50

Comment (sobolevn) [reply stubs/boltons/boltons/ecoutils.pyi:31]: In this case it is better to add it to `@tests/stubtest_allowlist.txt` with a comment.

File: stubs/boltons/boltons/ecoutils.pyi [hunk 7]
```diff
@@ -0,0 +1,34 @@
+# TODO: DONE!
+
+from typing import Any, NoReturn
+
+ECO_VERSION: str
+PY_GT_2: bool
+getrandbits: Any
```
[the remark this answers is not in the document]
Comment (AABur) [reply stubs/boltons/boltons/ecoutils.pyi]: It doesn't work that way. `stubtest` error
I need help!

File: stubs/boltons/boltons/ecoutils.pyi (hunk 6 above)
Comment (AABur) [reply stubs/boltons/boltons/ecoutils.pyi]: DONE!

**Comment by AlexWaygood** [comment]
@AABur, is this ready for a complete review, or do you just want help on the stubtest error for now? (I'm asking because this PR is still marked as a "draft PR" currently 🙂.)

**Comment by AABur** [comment]
> @AABur, is this PR ready for a complete review, or do you just want help on the stubtest error for now? (I'm asking because this PR is still marked as a "draft PR" currently 🙂.)

Ready for the complete review now!

**Comment by AABur** [comment]
@AlexWaygood sorry to bother you, but it's been a week. Any chance of my PR being accepted, or will there be any constructive comments?

**Comment by AlexWaygood** [comment]
Thanks for the ping @AABur. I haven't had a chance to have a look at this yet (we've got little bit of a backlog at the moment, and we're all volunteers!). I'll try to take a look soon 🙂

File: stubs/boltons/boltons/socketutils.pyi [hunk 8]
```diff
@@ -0,0 +1,74 @@
+import socket
+from _typeshed import Incomplete
+
+# class RLock:
+#     def __enter__(self) -> None: ...
+#     def __exit__(self, exctype, excinst, exctb) -> None: ...
+
```
Comment (AlexWaygood) [inline stubs/boltons/boltons/socketutils.pyi:7]: Why is this commented out?

```suggestion
```

File: stubs/boltons/boltons/statsutils.pyi [hunk 9]
```diff
@@ -0,0 +1,62 @@
+# TODO: DONE!
+
```
Comment (AlexWaygood) [inline stubs/boltons/boltons/statsutils.pyi:2]:
```suggestion
```

File: stubs/boltons/boltons/statsutils.pyi [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+    def trim_relative(self, amount: float = ...) -> NoReturn: ...
+    def get_histogram_counts(self, bins: int | None = ..., **kw) -> int: ...
+    def format_histogram(self, bins: int | None = ..., **kw) -> str: ...
+    def describe(self, quantiles: list[float] | None = ..., format: str | None = ...) -> dict[str, float] | list[float] | str: ...
+
+def describe(
+    data: list[float], quantiles: list[float] | None = ..., format: str | None = ...
+) -> dict[str, float] | list[float] | str: ...
+
+# func: Incomplete
```
Comment (AlexWaygood) [inline stubs/boltons/boltons/statsutils.pyi:47]:
```suggestion
```

File: stubs/boltons/boltons/strutils.pyi [hunk 11]
```diff
@@ -0,0 +1,66 @@
+# TODO: DONE!
+
```
Comment (AlexWaygood) [inline stubs/boltons/boltons/strutils.pyi:2]:
```suggestion
```

File: stubs/boltons/boltons/tbutils.pyi [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+    def tb_frame_str(self): ...
+
+class _DeferredLine:
+    filename: Incomplete
+    lineno: Incomplete
+    def __init__(self, filename, lineno, module_globals: Incomplete | None = ...) -> None: ...
+    def __eq__(self, other): ...
+    def __ne__(self, other): ...
+    def __len__(self): ...
+
```
Comment (AlexWaygood) [inline stubs/boltons/boltons/tbutils.pyi:29]: We don't usually include definitions for private things in typeshed

```suggestion
```

File: stubs/boltons/boltons/timeutils.pyi [hunk 13]
```diff
@@ -0,0 +1,65 @@
+# TODO: DONE!
+
```
Comment (AlexWaygood) [inline stubs/boltons/boltons/timeutils.pyi:2]:
```suggestion
```

File: stubs/boltons/boltons/urlutils.pyi [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+    def to_text(self, full_quote: bool = ...): ...
+    def __unicode__(self): ...
+    def __eq__(self, other): ...
+    def __ne__(self, other): ...
+
+class _sockaddr(ctypes.Structure): ...
+
+# WSAStringToAddressA: Incomplete
+# WSAAddressToStringA: Incomplete
+
```
Comment (AlexWaygood) [inline stubs/boltons/boltons/urlutils.pyi:76]:
```suggestion
```

**Review by AlexWaygood — DISMISSED** [review]
Thanks @AABur! This looks like a good start, but there's a few things that need fixing; see my comments below.

In general, I'm a bit confused by how much commented-out code there is. Either something exists at runtime (in which case it should be included in the stub), or it doesn't. There shouldn't really be much of a need, if any, to include large chunks of commented-out code in typeshed. If you _are_ including a large chunk of commented-out code, you need to explain _why_ it's commented-out, or it's very confusing.

Similarly, we don't really need the `# TODO: DONE` comments everywhere. `TODO` comments are for when there's something in a file that needs more work in the future :)

**Comment by AABur** [comment]
@AlexWaygood you are correct! I should have removed the comment-out code carefully before doing the PR
DONE!

File: stubs/boltons/boltons/cacheutils.pyi [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+xrange = range
+basestring: Any
+PREV: int
+NEXT: int
+KEY: int
+VALUE: int
+DEFAULT_MAX_SIZE: int
+
+class LRI(dict[_KT, _VT]):
+    hit_count: int
```
Comment (JelleZijlstra) [inline stubs/boltons/boltons/cacheutils.pyi:17]:
```suggestion
    hit_count: int
    miss_count: int
    soft_miss_count: int
```
These are documented in the class docstring

File: stubs/boltons/boltons/cacheutils.pyi [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+    max_size: int
+    on_miss: Callable[[_KT], _VT]
+    def __init__(self, max_size: int = ..., values: Incomplete | None = ..., on_miss: Incomplete | None = ...) -> NoReturn: ...
```
Comment (JelleZijlstra) [inline stubs/boltons/boltons/cacheutils.pyi:20]:
```suggestion
    def __init__(self, max_size: int = ..., values: Incomplete | None = ..., on_miss: Callable[[_KT], _VT] | None = ...) -> None: ...
```

File: stubs/boltons/boltons/cacheutils.pyi [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+    def __setitem__(self, key: _KT, value: _VT) -> NoReturn: ...
```
Comment (JelleZijlstra) [inline stubs/boltons/boltons/cacheutils.pyi:21]:
```suggestion
    def __setitem__(self, key: _KT, value: _VT) -> None: ...
```
`NoReturn` should be used if the function never returns (e.g. if it throws an exception), not if it simply returns None. This probably affects a bunch of other places; please fix them too.

File: stubs/boltons/boltons/cacheutils.pyi [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+    def __getitem__(self, key: _KT) -> _VT: ...
+    def get(self, key, default: Incomplete | None = ...): ...
+    def __delitem__(self, key: _KT) -> NoReturn: ...
+    def pop(self, key, default=...): ...
+    def popitem(self): ...
+    def clear(self) -> None: ...
+    def copy(self): ...
+    def setdefault(self, key, default: Incomplete | None = ...): ...
+    def update(self, E, **F) -> None: ...  # type: ignore
```
Comment (JelleZijlstra) [inline stubs/boltons/boltons/cacheutils.pyi:30]: E and F seem undefined, why is the error ignored?

File: stubs/boltons/boltons/cacheutils.pyi [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+    def __eq__(self, other): ...
+    def __ne__(self, other): ...
+
+class LRU(LRI[Any, Any]):
+    def __getitem__(self, key): ...
```
Comment (JelleZijlstra) [inline stubs/boltons/boltons/cacheutils.pyi:35]:
```suggestion
class LRU(LRI[_KT, _VT]):
    def __getitem__(self, key: _KT) -> _VT: ...
```
No reason not to keep this generic.

File: stubs/boltons/boltons/debugutils.pyi [hunk 20]
```diff
@@ -0,0 +1,5 @@
+from typing import NoReturn
+
+def pdb_on_signal(signalnum: int | None = ...) -> NoReturn: ...
+def pdb_on_exception(limit: int = ...) -> NoReturn: ...
```
Comment (JelleZijlstra) [inline stubs/boltons/boltons/debugutils.pyi:4]:
```suggestion
def pdb_on_signal(signalnum: int | None = ...) -> None: ...
def pdb_on_exception(limit: int = ...) -> None: ...
```

File: stubs/boltons/boltons/dictutils.pyi [hunk 21]
```diff
@@ -0,0 +1,87 @@
+from collections.abc import Generator, Hashable, ItemsView, Iterable, KeysView, ValuesView
+from typing import Any, NoReturn, TypeVar
+from typing_extensions import TypeAlias
+
+_KT = TypeVar("_KT")
+_VT = TypeVar("_VT")
+
+class OrderedMultiDict(dict[_KT, _VT]):
+    def add(self, k: _KT, v: _VT) -> NoReturn: ...
+    def addlist(self, k: _KT, v: Iterable[object]) -> NoReturn: ...
```
Comment (JelleZijlstra) [inline stubs/boltons/boltons/dictutils.pyi:10]:
```suggestion
    def addlist(self, k: _KT, v: Iterable[_VT]) -> None: ...
```

File: stubs/boltons/boltons/dictutils.pyi [hunk 22]
```diff
@@ -0,0 +1,87 @@
...
+    def clear(self) -> NoReturn: ...
+    def copy(self) -> OrderedMultiDict[_KT, _VT]: ...
```
Comment (JelleZijlstra) [inline stubs/boltons/boltons/dictutils.pyi:12]:
```suggestion
    def copy(self: Self) -> Self: ...
```
Using `from _typeshed import Self`

File: stubs/boltons/boltons/dictutils.pyi [hunk 23]
```diff
@@ -0,0 +1,87 @@
...
+    def counts(self) -> OrderedMultiDict[_KT, _VT]: ...
+    @classmethod
+    def fromkeys(cls, keys: _KT, default: _VT | None = ...) -> OrderedMultiDict[_KT, _VT]: ...  # type: ignore
```
Comment (JelleZijlstra) [inline stubs/boltons/boltons/dictutils.pyi:15]: Type ignores should generally have a comment explaining why they're there.

File: stubs/boltons/boltons/dictutils.pyi [hunk 24]
```diff
@@ -0,0 +1,87 @@
...
+    def todict(self, multi: bool = ...) -> dict[_KT, _VT]: ...
+    def update(self, E: dict[_KT, _VT] | Iterable[object], **F) -> NoReturn: ...  # type: ignore
+    def update_extend(self, E: dict[_KT, _VT] | Iterable[object], **F) -> NoReturn: ...
+    def values(self, multi: bool = ...) -> list[_VT]: ...  # type: ignore
+    def viewitems(self) -> ItemsView[_KT, _VT]: ...
+    def viewkeys(self) -> KeysView[_KT]: ...
+    def viewvalues(self) -> ValuesView[_VT]: ...
+
+OMD: TypeAlias = OrderedMultiDict[Any, Any]
+MultiDict: TypeAlias = OrderedMultiDict[Any, Any]
```
Comment (JelleZijlstra) [inline stubs/boltons/boltons/dictutils.pyi:39]:
```suggestion
OMD: TypeAlias = OrderedMultiDict[_KT, _VT]
MultiDict: TypeAlias = OrderedMultiDict[_KT, _VT]
```
Keep the aliases generic.

File: stubs/boltons/boltons/dictutils.pyi [hunk 25]
```diff
@@ -0,0 +1,87 @@
...
+
+class OneToOne(dict[_KT, _VT]):
+    inv: dict[_VT, _KT]
+    def clear(self) -> None: ...
+    def copy(self) -> OneToOne[_KT, _VT]: ...
+    def pop(self, key: _KT, default: _VT = ...) -> _VT: ...  # type: ignore
+    def popitem(self) -> tuple[_KT, _VT]: ...
+    def setdefault(self, key: _KT, default: _VT | None = ...) -> _VT: ...
+    @classmethod
+    def unique(cls, *a, **kw) -> NoReturn: ...
```
Comment (JelleZijlstra) [inline stubs/boltons/boltons/dictutils.pyi:53]:
```suggestion
    def unique(cls: type[Self], *a, **kw) -> Self: ...
```

File: stubs/boltons/boltons/dictutils.pyi [hunk 26]
```diff
@@ -0,0 +1,87 @@
...
+    def __len__(self): ...
+    def __setitem__(self, key, vals) -> None: ...
+    def add(self, key: Hashable, val: object) -> NoReturn: ...
+    def get(self, key: Hashable, default: frozenset[object] = ...) -> object: ...
+    def iteritems(self) -> Generator[tuple[Hashable, object], None, None]: ...
+    def keys(self): ...  # incomplete
+    def remove(self, key: Hashable, val: object) -> NoReturn: ...
+    def replace(self, key: Hashable, newkey: Hashable) -> NoReturn: ...
+    def update(self, iterable) -> None: ...
+    inv: tuple[object]
```
Comment (JelleZijlstra) [inline stubs/boltons/boltons/dictutils.pyi:73]:
```suggestion
    inv: ManyToMany
```

File: stubs/boltons/boltons/dictutils.pyi [hunk 27]
```diff
@@ -0,0 +1,87 @@
...
+    def clear(self) -> None: ...
+    def copy(self) -> OneToOne[_KT, _VT]: ...
+    def pop(self, key: _KT, default: _VT = ...) -> _VT: ...  # type: ignore
+    def popitem(self) -> tuple[_KT, _VT]: ...
+    def setdefault(self, key: _KT, default: _VT | None = ...) -> _VT: ...
+    @classmethod
+    def unique(cls, *a, **kw) -> NoReturn: ...
+    def update(self, dict_or_iterable: dict[_KT, _VT] | Iterable[object], **kw) -> NoReturn: ...  # type: ignore
+
+class ManyToMany:
```
Comment (JelleZijlstra) [inline stubs/boltons/boltons/dictutils.pyi:56]: Should probably be generic.

File: stubs/boltons/boltons/dictutils.pyi [hunk 28]
```diff
@@ -0,0 +1,87 @@
...
+    def add(self, key: Hashable, val: object) -> NoReturn: ...
+    def get(self, key: Hashable, default: frozenset[object] = ...) -> object: ...
+    def iteritems(self) -> Generator[tuple[Hashable, object], None, None]: ...
+    def keys(self): ...  # incomplete
+    def remove(self, key: Hashable, val: object) -> NoReturn: ...
+    def replace(self, key: Hashable, newkey: Hashable) -> NoReturn: ...
+    def update(self, iterable) -> None: ...
+    inv: tuple[object]
+
+def subdict(d: dict[_KT, _VT], keep: Iterable[_VT] | None = ..., drop: Iterable[_VT] | None = ...) -> dict[_KT, _VT]: ...
```
Comment (JelleZijlstra) [inline stubs/boltons/boltons/dictutils.pyi:75]:
```suggestion
def subdict(d: dict[_KT, _VT], keep: Iterable[_KT] | None = ..., drop: Iterable[_KT] | None = ...) -> dict[_KT, _VT]: ...
```
keep and drop are about keys, not values

File: stubs/boltons/boltons/dictutils.pyi [hunk 29]
```diff
@@ -0,0 +1,87 @@
...
+
+class FrozenDict(dict[_KT, _VT]):
+    def __copy__(self): ...
+    def clear(self, *a, **kw) -> NoReturn: ...
+    @classmethod
+    def fromkeys(cls, keys: Iterable[_KT], value: _VT | None = ...) -> FrozenDict[_KT, _VT]: ...  # type: ignore
+    def pop(self, *a, **kw) -> NoReturn: ...
+    def popitem(self, *a, **kw) -> NoReturn: ...
+    def setdefault(self, *a, **kw) -> NoReturn: ...
+    def updated(self, *a, **kw) -> FrozenDict[_KT, _VT]: ...
```
Comment (JelleZijlstra) [inline stubs/boltons/boltons/dictutils.pyi:87]:
```suggestion
    def updated(self: Self, *a, **kw) -> Self: ...
```

File: stubs/boltons/boltons/easterutils.pyi [hunk 30]
```diff
@@ -0,0 +1,3 @@
+from typing import NoReturn
+
+def gobs_program() -> NoReturn: ...
```
Comment (JelleZijlstra) [inline stubs/boltons/boltons/easterutils.pyi:3]: Here `NoReturn` is actually correct

**Review by JelleZijlstra — COMMENTED** [review]
Only looked at files up to `excutils`.

File: stubs/boltons/boltons/cacheutils.pyi (hunk 18 above)
Comment (AABur) [reply stubs/boltons/boltons/cacheutils.pyi:30]: I just don't know how to type it

File: stubs/boltons/boltons/easterutils.pyi (hunk 30 above)
Comment (AABur) [reply stubs/boltons/boltons/easterutils.pyi:3]: Thank you for the clarification on NoReturn and None

**Comment by AABur** [comment]
@JelleZijlstra thanks for the comments. Everything is clear. I will update PR soon

**Comment by AABur** [comment]
> Ping @AABur, have you had a chance to take a look at Jelle's review? :)

Yes, I did look at Jelle's review earlier, but I got busy with other tasks afterwards. I'll need to refresh my memory on the specifics before discussing it further. :)
I am now ready to return to this task. It will take a few days

**Comment by AlexWaygood** [comment]
`Callable[Any, Any]` is invalid. If you want to say "A callable that accepts a single parameter, which could be any value", you want `Callable[[Any], Any]`. If you want to say "A callable with any arbitrary number of parameters, any of which could be any value", you want `Callable[..., Any]`.

File: stubs/boltons/boltons/cacheutils.pyi (hunk 18 above)
Comment (AABur) [reply stubs/boltons/boltons/cacheutils.pyi:30]: I have no idea how to annotate it correctly.

**Comment by AABur** [comment]
Did what I could based on the knowledge and skills available. I need help for further improvement.
Where it contains `# type: ignore` - it means that it did not succeed to annotate properly. I will change it to `Incomplete` sometime in the future.

File: stubs/boltons/boltons/cacheutils.pyi (hunk 18 above)
Comment (JelleZijlstra) [reply stubs/boltons/boltons/cacheutils.pyi:32]: Then we should just annotate it with `*args`/`**kwargs` for now. I want to avoid type ignores.

**Comment by JelleZijlstra** [comment]
I went over the code locally and made a number of changes:

- Target `boltons 23.0.0`, which has now been released
- Remove various compatibility aliases (e.g., `unicode`) that shouldn't be used directly
- Fix many incorrect uses of `NoReturn`
- Fix various other incorrect types
- Add some missing types

Stubtest will probably have some complaints

**Review by JelleZijlstra — APPROVED** [review]
There may be some more issues, but I think I caught most of them.

@AlexWaygood do you want to take another look?

**Comment by AlexWaygood** [comment]
> There may be some more issues, but I think I caught most of them.
> 
> @AlexWaygood do you want to take another look?

If you're happy, I'm happy :)

Let's merge this now

**Comment by sobolevn** [comment]
Congrats @AABur 🎉

**Comment by sobolevn** [comment]
@kyawgyi99996 what kind of help do you need?

**Pull request merged**